### PR TITLE
Allow user to update positions of DynamicMesh

### PR DESCRIPTION
**Pull request by alteous**
This PR allows the user to alter the vertices of a `DynamicMesh` at runtime.

### CODE REVIEW

File: src/mesh.rs [hunk 1]
```diff
@@ -87,6 +88,16 @@ impl Mesh {
 }
 
 impl DynamicMesh {
+    /// Returns the indexed vertex position of the geometry base shape.
+    pub fn vertex(&self, index: usize) -> mint::Point3<f32> {
```
Comment (vitvakatu) [inline src/mesh.rs]: Where is guarantee that we always can return value at any index? Need to use `Option` or smth else

File: src/factory/mod.rs [hunk 2]
```diff
@@ -468,7 +485,44 @@ impl Factory {
...
+            TANGENT_X
+        };
+
+        let uv = if let Some(uv) = mesh.geometry.base_shape.tex_coords.get(index) {
+            uv.clone().into()
+        } else {
+            [0.0, 0.0]
+        };
+
+        mesh.geometry.base_shape.vertices[index] = position;
```
Comment (vitvakatu) [inline src/factory/mod.rs]: index can be invalid, need to check

File: examples/mesh-update.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
+extern crate cgmath;
```
Comment (vitvakatu) [inline examples/mesh-update.rs]: This example is too complicated. I would like to see either smaller one, or none at all.

**Comment by vitvakatu** [comment]
By the way - it would be better to expose some way to access [`Writer`](https://docs.rs/gfx/0.16.1/gfx/mapping/struct.Writer.html) directly, because acquiring it for every single vertex isn't good.

File: src/mesh.rs (hunk 1 above)
Comment (alteous) [reply src/mesh.rs]: Agreed.

File: examples/mesh-update.rs (hunk 3 above)
Comment (alteous) [reply examples/mesh-update.rs]: Yes, I agree. I'll re-implement with a tetrahedron instead of a cube.

File: src/factory/mod.rs (hunk 2 above)
Comment (alteous) [reply src/factory/mod.rs]: What would we do if the index is out of range?

**Comment by alteous** [comment]
> it would be better to expose some way to access Writer directly

We have to be careful not to expose the internal `Vertex` type. How about the following?

```rust
pub fn overwrite_positions(
    &mut self,
    mesh: &mut DynamicMesh,
    updates: &[(usize, mint::Point3<f32>)],
);
```

**Comment by vitvakatu** [comment]
Looks good as for me.

File: examples/mesh-update.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    let mut controls = three::controls::Orbit::builder(&cam)
+        .position([0.0, 2.0, -5.0])
+        .target([0.0, 0.0, 0.0])
+        .build();
+
+    let mut dir_light = win.factory.directional_light(0xFFFFFF, 0.9);
+    dir_light.look_at([15.0, 35.0, 35.0], [0.0, 0.0, 2.0], None);
+    win.scene.add(&dir_light);
+
+    let geometry = make_tetrahedron_geometry();
```
Comment (alteous) [inline examples/mesh-update.rs]: Perhaps we should add a `Geometry::tetrahedron` constructor just for this example?

File: src/factory/mod.rs [hunk 5]
```diff
@@ -468,7 +485,49 @@ impl Factory {
...
+    ///
+    /// Panics if any of the indices exceed the number of vertices in the
+    /// mesh geometry.
+    pub fn update_positions(
+        &mut self,
+        mesh: &mut DynamicMesh,
+        updates: &[(usize, mint::Point3<f32>)],
+    ) {
+        self.hub.lock().unwrap().update_mesh(mesh);
+        let mut mapping = self.backend.write_mapping(&mesh.dynamic.buffer).unwrap();
```
Comment (alteous) [inline src/factory/mod.rs]: Only take one lock now.

File: src/factory/mod.rs [hunk 6]
```diff
@@ -468,7 +485,49 @@ impl Factory {
...
+            };
+
+            let uv = if let Some(uv) = mesh.geometry.base_shape.tex_coords.get(index) {
+                uv.clone().into()
+            } else {
+                [0.0, 0.0]
+            };
+
+            mesh.geometry.base_shape.vertices[index] = position;
+            mapping[index] = Vertex {
```
Comment (alteous) [inline src/factory/mod.rs]: The possibility of panic is noted in the docs.

File: src/factory/mod.rs [hunk 7]
```diff
@@ -468,7 +485,49 @@ impl Factory {
...
     }
 
-    /// Update the geometry of `DynamicMesh`.
+    /// Update the vertices of the geometry of a `DynamicMesh`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any of the indices exceed the number of vertices in the
+    /// mesh geometry.
+    pub fn update_positions(
```
Comment (alteous) [inline src/factory/mod.rs]: Renamed from `overwrite_position`.

File: examples/mesh-update.rs (hunk 4 above)
Comment (vitvakatu) [reply examples/mesh-update.rs]: Not sure, tetrahedron isn't best shape for everyday usage.
Example became much nicer, thanks!

File: src/mesh.rs [hunk 8]
```diff
@@ -87,6 +88,17 @@ impl Mesh {
 }
 
 impl DynamicMesh {
+    /// Returns the indexed vertex position of the geometry base shape,
+    /// or `None` if the index is out of bounds.
+    pub fn vertex(&self, index: usize) -> Option<mint::Point3<f32>> {
+        self.geometry.base_shape.vertices.get(index).map(Clone::clone)
```
Comment (vitvakatu) [inline src/mesh.rs]: `mint` types implement `Copy`, don't they?

Comment (alteous) [reply src/mesh.rs]: Yes, but `Option` does not.

**Comment by alteous** [comment]
Update: I've renamed the function `update_mesh_vertices` which now takes an `Iterator` instead of a slice.

**Comment by alteous** [comment]
I would like to hear @kvark 's opinion too.

**Comment by alteous** [comment]
Thinking ahead, suppose we have two separate functions `update_mesh_positions` and `update_mesh_normals`. If the user required both then we would have to map the buffer twice, which isn't ideal.

Perhaps later we should try another strategy. I'm thinking along the lines of.

```rust
factory
    .update_mesh()
    .write_vertices(iter::once((123, [0.0, 0.0, 0.0].into()))
    .write_normals(iter::once((456, [0.0, 1.0, 0.0].into())
    .exec();
```

File: src/mesh.rs (hunk 8 above)
Comment (vitvakatu) [reply src/mesh.rs]: Option has `cloned()` afaik, it looks much nicer

Comment (alteous) [reply src/mesh.rs]: TIL, thanks. :smile:

File: examples/mesh-update.rs [hunk 9]
```diff
@@ -0,0 +1,71 @@
+extern crate cgmath;
+extern crate three;
+extern crate mint;
+
+use cgmath::prelude::*;
+use std::iter;
+use std::f32::consts::PI;
+
+fn make_tetrahedron_geometry() -> three::Geometry {
+    let vertices = vec![
```
Comment (kvark) [inline examples/mesh-update.rs]: nice, I ❤️ tetrahedrons :)

File: src/factory/mod.rs [hunk 10]
```diff
@@ -255,12 +255,29 @@ impl Factory {
...
+                .create_upload_buffer(data.len())
                 .unwrap();
-            (data.len(), buf)
+            // TODO: Workaround for not having a 'write-to-slice' capability.
+            // Reason: The renderer copies the entire staging buffer upon updates.
+            {
+                let mut mapping = self.backend
+                    .write_mapping(&upload_buf)
+                    .unwrap();
+                for (i, vertex) in data.iter().cloned().enumerate() {
```
Comment (kvark) [inline src/factory/mod.rs]: how about this:
```rust
self.backend
    .write_mapping(&upload_buf)
    .unwrap()
    .copy_from_slice(&data);
```

File: src/factory/mod.rs [hunk 11]
```diff
@@ -468,7 +485,26 @@ impl Factory {
...
     }
 
-    /// Update the geometry of `DynamicMesh`.
+    /// Update the vertex positions of the geometry of a `DynamicMesh`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any of the indices exceed the number of vertices in the
+    /// mesh geometry.
+    pub fn update_mesh_vertices<I: Iterator<Item = (usize, mint::Point3<f32>)>>(
```
Comment (kvark) [inline src/factory/mod.rs]: I think we should just provide a mapping API for dynamic meshes, e.g.
```rust
{
    let mut mapping = factory.map_vertices(&mesh);
    for i in 0.. 5 {
         mapping[i] = MyVertex {..};
    }
}
```
This mapping object would implement `DerefMut` to our vertex format and hold `factory` by a mutable reference, then issue an actual update message upon destruction (a message would make the renderer to copy the data from upload -> data buffer upon the first use or something).

File: src/mesh.rs (hunk 8 above)
Comment (kvark) [reply src/mesh.rs]: yep, let's just use `cloned` here

File: src/factory/mod.rs (hunk 11 above)
Comment (alteous) [reply src/factory/mod.rs]: This would be a much simpler approach. The downside is having to expose out vertex format.

**Comment by alteous** [comment]
Should be ready to merge now.

File: src/factory/mod.rs [hunk 12]
```diff
@@ -65,6 +65,9 @@ const QUAD: [Vertex; 4] = [
     },
 ];
 
+/// Mapping writer.
+pub type MapVertices<'a> = gfx::mapping::Writer<'a, BackendResources, Vertex>;
```
Comment (kvark) [inline src/factory/mod.rs]: IIRC, the only place we currently expose gfx-rs is custom pipelines, which to be fair could be hidden behind a feature flag just fine. You proposed a wrapper before, I think that was a good idea, totally fine for a follow-up.

File: src/factory/mod.rs [hunk 13]
```diff
@@ -275,12 +278,21 @@ impl Factory {
...
             let data = Self::mesh_vertices(&geometry.base_shape);
-            let buf = self.backend
+            let dest_buf = self.backend
                 .create_buffer_immutable(&data, gfx::buffer::Role::Vertex, gfx::memory::TRANSFER_DST)
                 .unwrap();
-            (data.len(), buf)
+            let upload_buf = self.backend.create_upload_buffer(data.len()).unwrap();
+            // TODO: Workaround for not having a 'write-to-slice' capability.
+            // Reason: The renderer copies the entire staging buffer upon updates.
+            {
```
Comment (kvark) [inline src/factory/mod.rs]: I don't think you need the scope here?

**Pull request merged**